discovery.go

```diff
@@ -0,0 +1,225 @@
+package pubsub
+
+import (
+	"context"
+	"github.com/libp2p/go-libp2p-core/discovery"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"sync"
+	"time"
+)
+
+// discover represents the discovery pipeline.
+// The discovery pipeline handles advertising and discovery of peers
+type discover struct {
+	p *PubSub
+
+	// discovery assists in discovering and advertising peers for a topic
+	discovery discovery.Discovery
+
+	// advertising tracks which topics are being advertised
+	advertising map[string]context.CancelFunc
+
+	// bootstrapped tracks which topics have been bootstrapped
+	bootstrapped sync.Map
+
+	// discover handles continuing peer discovery
+	discover chan *Discover
+
+	// ongoingDiscovery tracks ongoing discovery requests
+	ongoingDiscovery map[string]struct{}
+
+	// doneDiscovery handles completion of a discovery request
+	doneDiscovery chan string
+}
+
+// newValidation creates a new validation pipeline
+func newDiscover() *discover {
+	return &discover{
+		bootstrapped:     sync.Map{},
+		discover:         make(chan *Discover),
+		ongoingDiscovery: make(map[string]struct{}),
+		doneDiscovery:    make(chan string),
+		advertising:      make(map[string]context.CancelFunc),
+	}
+}
+
+// Start attaches the discovery pipeline to a pubsub instance and starts event loop
+func (d *discover) Start(p *PubSub) {
+	d.p = p
+	go d.discoverLoop()
+}
+
+func (d *discover) discoverLoop() {
+	for {
+		select {
+		case discover := <-d.discover:
+			topic := discover.topic
+			if _, ok := d.ongoingDiscovery[topic]; !ok {
+				d.ongoingDiscovery[topic] = struct{}{}
+				go func() {
+					d.handleDiscovery(d.p.ctx, topic, discover.opts)
+					d.doneDiscovery <- topic
+				}()
+			}
+		case topic := <-d.doneDiscovery:
+			delete(d.ongoingDiscovery, topic)
+		case <-d.p.ctx.Done():
+			log.Info("pubsub discoverloop shutting down")
+			return
+		}
+	}
+}
+
+// advertise advertises this node's interest in a topic to a discovery service
+func (d *discover) advertise(topic string) {
+	if d.discovery != nil {
```

Comment:
why this check? It shouldn't start at all if the `discovery` object is `nil`.

Reply:
It's just a question of where to put the `nil` checks. Either I put them in the `discover` functions, or I put them in pubsub (e.g. in `handleAddSubscription`) I opted to remove any extra lines of code from pubsub and put them here instead.

If you disagree I can move it, just let me know.

Reply:
ok